[WellSQL Migration] Migrate `WCNewVisitorStatsModel` to Room

ParaskP7 opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

Closes: [AINFRA-554](https://linear.app/a8c/issue/AINFRA-554)

---

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR migrates `WCNewVisitorStatsModel` from [WellSQL](https://github.com/wordpress-mobile/wellsql) to [Room](https://developer.android.com/training/data-storage/room) database.

---

### Reviewers

In addition to having @wzieba as the main "code" reviewer on this change, I added `@woocommerce/android-developers` to randomly assign a product engineers as an "extra" review (~~@irfano~~ @hichamboushaba), primarily because, both on `trunk` and this PR branch, I've these questions (⚠️):

1. `OrderStatsRestClient.fetchNewVisitorStats(...)`: This is where the only place where `WCNewVisitorStatsModel` get instantiated and its `isCustomField` is being driven by `startDate != null`, questions:
    - Why is `startDate` driving this `isCustomField` field? 🤔
    - Due to the above, during my testing, `isCustomField` was always `1`, or else `true`, just because `startDate` was always non-null, no matter what option the user chooses from the UI (`My store` -> `Performance` -> `Calendar` icon). Is that expected? 🤔
2. `WCNewVisitorStatsModel`: During my testing, when using `Custom` from the UI (`My store` -> `Performance` -> `Calendar` icon), I noticed that the `Visitors` (and `Conversion`) data is unavailable for any date range picked other than a specific "single" day. Is that expected? 🤔

---

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

#### `My store` + `Performance`

1. Navigate on the `My store` tab and verify that the `Performance` section is being correctly populated with the expected number of `Visitors`.
2. On that `Performance` section, click on the `Calendar` icon and select `Today`. Verify that the section's title changes to something like `Today Sep 9, 2025` and that the number of `Visitors` is updated accordingly.
3. Similarly, select `This Week`, `This Month` and `This Year` and verify everything is working as expected.
4. Open `App Inspection`, check the `NewVisitorStatsEntity` table, and notice (⚠️) that `isCustomField` is `1`.
5. Now, select `Custom` and edit the section's title picking a specific "single" day like `Sep 9 - Sep 9`, click `SAVE`. Verify that the section's title changes to something like `Custom Sep 9, 2025 - Sep 9, 2025` and that the number of `Visitors` is updated accordingly.
6. Then, edit the section's title again, picking any other range, except for a "single" day, like `Sep 8 - Sep 9`, click `SAVE`. Verify that the section's title changes to something like `Custom Sep 8, 2025 - Sep 9, 2025` and notice (⚠️) that the number of `Visitors` (and `Conversion`) is actually **unavailable**. (see also `Reviewers` related questions).
7. Open `App Inspection`, check the `NewVisitorStatsEntity` table, and verify that `isCustomField` is `1`.

---

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

<img width="1915" height="744" alt="image" src="https://github.com/user-attachments/assets/d18c7b6d-20c2-496f-b5da-8cfa04183b8a" />

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

ParaskP7 committed:
Refactor: Optimize imports on wc stats store

ParaskP7 committed:
Analysis: Replace order stats error type with enum entries

Warning: "'Enum.values()' is recommended to be replaced by
'Enum.entries' since 1.9"

ParaskP7 committed:
Analysis: Fix unreachable code on get new visitor stats

Warning: "Unreachable code"

ParaskP7 committed:
Analysis: Merge call chain to associate on get xyz stats

Warning: "Call chain on a collection type may be simplified"

ParaskP7 committed:
Refactor: Reformat wc stats store (test)

ParaskP7 committed:
Refactor: Make getNewVisitorStats suspend functions

FYI: This is done in preparation to the subsequent WellSQL to Room
table migration.

ParaskP7 committed:
Refactor: Optimize imports on wc stats store test (and fix typo)

ParaskP7 committed:
Test: Move wc stats test utils and wc stats store test to fluxc-plugin

FYI: This is done in preparation to the subsequent WellSQL to Room
table migration.

ParaskP7 committed:
Cleanup: Remove unnecessary rows affected from on wc stats changed

FYI: This is done in preparation to the subsequent WellSQL to Room
table migration.

ParaskP7 committed:
Migrate: 'WCNewVisitorStatsModel' from wellsql to room

FYI: This change includes but is not limited to the:
- Entity Introduction
- DAO Introduction (With Test)
- WCAndroidDatabase Migration
- Code Adjustments (Store)
- Test Adjustments (Store)

ParaskP7 committed:
Db: Drop 'WCNewVisitorStatsModel' table from 'wp-fluxc'

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/NewVisitorStatsDao.kt:
```diff
@@ -0,0 +1,66 @@
+import org.wordpress.android.fluxc.store.WCStatsStore.StatsGranularity
+
+@Dao
+internal abstract class NewVisitorStatsDao {
```
Nah, it's okay if we keep this as `abstract class` instead, just like done elsewhere, thanks though! 🙇

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -208,19 +209,20 @@ class WCStatsStore @Inject constructor(
             }
-        } ?: return mapOf()
+        }
+        return mapOf()
```
You're right, and that's exactly what I did, actually a bit confused by this comment of yours... 🤔

wzieba replied:
Yes, the comment is incorrect - I guess LLM wrongly interpreted the inner return (line 203). 

A suggestion I have is to refactor for a single `return` point, but that's not critical.

```
val visitorMap = rawStats?.let { visitorStatsModel ->
        val periodIndex = visitorStatsModel.getIndexForField(WCNewVisitorStatsModel.VisitorStatsField.PERIOD)
        val fieldIndex = visitorStatsModel.getIndexForField(WCNewVisitorStatsModel.VisitorStatsField.VISITORS)

        if (periodIndex != -1 && fieldIndex != -1) {
            getVisitorsMap(
                periodIndex = periodIndex,
                fieldIndex = fieldIndex,
                dataList = visitorStatsModel.dataList
            )
        } else {
            emptyMap()
        }
    }

    return visitorMap ?: emptyMap()
```

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -234,7 +236,8 @@ class WCStatsStore @Inject constructor(
                 )
             }
-        } ?: return mapOf()
+        }
+        return mapOf()
```
[SAME](https://github.com/woocommerce/woocommerce-android/pull/14592#discussion_r2330711548)

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into migrate/wc-new-visitor-stats-model-to-room

wzieba commented on config/detekt/baseline.xml:
```diff
@@ -2271,21 +2271,6 @@
     <ID>UnitTestNamingRule:WCShippingLabelStoreTest.kt$WCShippingLabelStoreTest$@Test fun `verify shipping address`()</ID>
     <ID>UnitTestNamingRule:WCStatsSqlUtilsTest.kt$WCStatsSqlUtilsTest$@Test @Suppress("LongMethod") fun testGetRawRevenueStatsForSiteAndUnit()</ID>
     <ID>UnitTestNamingRule:WCStatsSqlUtilsTest.kt$WCStatsSqlUtilsTest$@Test @Suppress("LongMethod") fun testSimpleInsertionAndRetrievalOfRevenueStats()</ID>
-    <ID>UnitTestNamingRule:WCStatsStoreTest.kt$WCStatsStoreTest$@Suppress("LongMethod") @Test fun testGetQuantityForMonths()</ID>
```
Nice work on this (here and previous PRs) - a good reminder to look at the `baseline` (Detekt or Lint) when working on technical debt 🏅

irfano commented:
I unassigned myself. Since I’ll be AFK until September 19, can you take a look at this, @JorgeMucientes or @hichamboushaba?

ParaskP7 commented:
> I unassigned myself. Since I’ll be AFK until September 19...

Oh, thanks @irfano and apologies, I (or algorithm) wasn't aware you're on AFK. 🫣

FYI: I am not sure if you had the change to use this `AFK/Github Busy Status Sync` (paaHJt-8Ri-p2) yet, it might help you folks in situations like this in the future. Cc @woocommerce/android-developers for further awareness. 🙏

> ...can you take a look at this, @JorgeMucientes or @hichamboushaba?

Since I've assign @JorgeMucientes on this previous #14520 PR of mine already, I'll go ahead and assign you @hichamboushaba on this one, let me know how you feel and I'll find an alternative. 🙏

wzieba reviewed (approved):
LGTM! 

I don't know the answers on the domain of the Stats, so I'll leave this to the product team. I also couldn't perform detailed tests due to log in issues (internal details: p1757495577372719-slack-CGPNUU63E ) but I don't want to keep this PR on hold because of this.

ParaskP7 commented:
Thanks for the review @wzieba (🙇) , I'll wait for another review from the product team, plus maybe you being able to successfully test this afterwards, and I'll then proceed with applying any of the review suggestions (like [this](https://github.com/woocommerce/woocommerce-android/pull/14592#discussion_r2333189583) one). 👍

hichamboushaba commented:
> I'll go ahead and assign you @hichamboushaba on this one, let me know how you feel and I'll find an alternative. 🙏

Sounds good @ParaskP7, but I hope it's fine not to review this until tomorrow 🙂

ParaskP7 commented:
> Sounds good @ParaskP7, but I hope it's fine not to review this until tomorrow 🙂

Totally fine @hichamboushaba , many thanks + appreciated, and no rush, no rush at all! 🙇 ❤️

hichamboushaba reviewed (approved):
Nice work @ParaskP7 👏, this works as expected, and I'll try to answer your questions below:

> OrderStatsRestClient.fetchNewVisitorStats(...): This is where the only place where WCNewVisitorStatsModel get instantiated and its isCustomField is being driven by startDate != null, questions:
> - Why is startDate driving this isCustomField field? 🤔
> - Due to the above, during my testing, isCustomField was always 1, or else true, just because startDate was always non-null, no matter what option the user chooses from the UI (My store -> Performance -> Calendar icon). Is that expected? 🤔

I also didn't know, but after looking at the history of this field, I understand now, back in 2019 when this table was [added](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1336), the app supported a "custom" date option in stats graph (that option was removed later), and the way the team worked on this was that they treated the `granularity` option as a date (so `StatsGranularity.DAYS` was equivalent to "Today") unless the `startDate` was specified, `startDate` was specified only for the `custom` date option in the graph.
This was confusing, and we [got rid](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2968) of this when we brought back the "Custom" date option last year, and we changed the logic so that `startDate` was always required.
I overlooked the column `isCustomField` of the DB when I worked on this, otherwise I should have deleted it as it's always `true` now.
Now whether you should delete it or not, I'll leave it up to you to decide this 🙂.

> WCNewVisitorStatsModel: During my testing, when using Custom from the UI (My store -> Performance -> Calendar icon), I noticed that the Visitors (and Conversion) data is unavailable for any date range picked other than a specific "single" day. Is that expected? 🤔

Yes, it's expected, but this has nothing to do with this migration, the `Visitors` count that's shown on the graph when no entry is selected is driven by the `VisitorSummaryStatsEntity`, and this used Room since it was [added](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2983).

----
Please let me know if you have any other questions, but the main point here is that this is ready for merge unless you want to delete the `isCustomField` column.

ParaskP7 commented:
👋 @hichamboushaba and thanks for reviewing/testing this, as well as answering my questions, much appreciated! 🙇 ❤️

> Nice work @ParaskP7 👏, this works as expected, and I'll try to answer your questions below:

Thank you for digging as deep to answer my question! 🙏 

> Please let me know if you have any other questions, but the main point here is that this is ready for merge unless you want to delete the isCustomField column.

Yea, I think it might be better (an opportunity) for us to delete this unused `isCustomField` column (should be hard), then make this whole solution, store/rest/persistence that bit less complicated. And if not for anything else, just to avoid having this discussion again in the future when someone else will try and touch this logic.

WILL DO 🏃 💨💨💨

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into migrate/wc-new-visitor-stats-model-to-room

# Conflicts:
#	libs/fluxc-plugin/schemas/org.wordpress.android.fluxc.persistence.WCAndroidDatabase/64.json

ParaskP7 committed:
Migrate: 'WCNewVisitorStatsModel' from wellsql to room (v2)

Due to 146ebde48c24bb0e934a5cb2af97ba990705b6f5 a merge conflict
appeared which got solved via 781b01b867c0f7f38b397eb40bd0d24e69696410,
with the main '64.json' file, but also a hidden merge conflict on
'WCAndroidDatabase', which includes:

- const val WC_DATABASE_VERSION = 65
- AutoMigration(from = 64, to = 65),

I accepted the current version of '64.json' and created a new one, the
'65.json' because the database version for this exact change has now
been updated to '65' (from '64').

ParaskP7 committed:
Cleanup: Remove unnecessary is custom field logic from new visitor stats

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
14592#pullrequestreview-3212230060

ParaskP7 committed:
Cleanup: Remove leftover custom keywork from new visitor stats dao

ParaskP7 committed:
Refactor: Rename insert/update to replace stat on new visitor stats dao

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -186,16 +184,13 @@ class WCStatsStore @Inject constructor(
-        val rawStats = WCVisitorStatsSqlUtils.getNewRawVisitorStatsForSiteGranularityQuantityAndDate(
-            site, granularity, quantity, date, isCustomField
-        )
+        val rawStats = newVisitorStatsDao.getStat(site.localId(), granularity, quantity, date)
```
Not sure what you're saying Mr. bot, the calling `getNewVisitorStats(...)` is itself a `suspend` function... 🤷

Ah, you're worried about the `context`, okay, got it, but ignoring, not necessary. 👍

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -208,19 +203,20 @@ class WCStatsStore @Inject constructor(
         val date = DateUtils.getDateTimeForSite(site, DATE_FORMAT_DAY, endDate)
-        val rawStats = WCVisitorStatsSqlUtils.getNewRawVisitorStatsForSiteGranularityQuantityAndDate(
-            site, granularity, quantity.toString(), date, true
+        val rawStats = newVisitorStatsDao.getStat(
+            site.localId(), granularity, quantity.toString(), date
```
[SAME](https://github.com/woocommerce/woocommerce-android/pull/14592#discussion_r2344143253)

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -337,29 +334,27 @@ class WCStatsStore @Inject constructor(
                 }
             } else {
-                val rowsAffected = WCVisitorStatsSqlUtils.insertOrUpdateNewVisitorStats(result.stats)
+                newVisitorStatsDao.replaceStat(result.stats)
```
[SAME](https://github.com/woocommerce/woocommerce-android/pull/14592#discussion_r2344143253)

hichamboushaba reviewed (approved):
Nice work @ParaskP7, thanks so much for the code cleanup
